sw/device/lib/base/bitfield.h

```diff
@@ -13,58 +14,101 @@ extern "C" {
+ * zero-bit is in the return value's zero-bit.
  *
  * @param bitfield Bitfield to get the field from.
  * @param field Field to read out from.
- * @return The read value, zero-extended to 32 bits.
+ * @return Zero-extended `field` from `bitfield`.
  */
-inline uint32_t bitfield_get_field32(uint32_t bitfield,
-                                     bitfield_field32_t field) {
+inline uint32_t bitfield_field32_read(uint32_t bitfield,
+                                      bitfield_field32_t field) {
   return (bitfield >> field.index) & field.mask;
 }
 
 /**
- * Sets `field` in `bitfield`.
+ * Writes `value` to `field` in `bitfield`.
  *
- * This function uses the `field` parameter to set a value
- * at a given offset in `bitfield`. The relevant portion of `bitfield`
- * is zeroed before the value is set.
+ * This function uses the `field` parameter to set specific bits in `bitfield`.
+ * The relevant portion of `bitfield` is zeroed before the bits are set to
+ * `value`.
  *
  * @param bitfield Bitfield to set the field in.
- * @param field Field within selected register field to be set.
- * @return The 32-bit resulting bitfield.
+ * @param field Field within bitfield to be set.
+ * @param value Value for the new field.
+ * @return `bitfield` with `field` set to `value`.
  */
-inline uint32_t bitfield_set_field32(uint32_t bitfield,
-                                     bitfield_field32_t field) {
+inline uint32_t bitfield_field32_write(uint32_t bitfield,
+                                       bitfield_field32_t field,
+                                       uint32_t value) {
   bitfield &= ~(field.mask << field.index);
-  bitfield |= (field.value & field.mask) << field.index;
+  bitfield |= (value & field.mask) << field.index;
+  return bitfield;
+}
+
+/**
+ * A single bit in a 32-bit bitfield.
+ *
+ * This denotes the position of a single bit, counting from the zero-bit.
+ *
+ * For instance, `(bitfield_bit_index_t)4` denotes the X-marked bit in the
+ * following 32-bit bitfield:
+ *
+ *     field:  0b--------'--------'--------'---X----
+ *     index:   31                                 0
+ *
+ * Restrictions: The value must not be greater than 31.
+ */
+typedef uint32_t bitfield_bit32_index_t;
+
+/**
+ * Reads the `bit_index`th bit in `bitfield`.
+ *
+ * @param bitfield Bitfield to get the bit from.
+ * @param bit_index Bit to read.
+ * @return `true` if the bit was one, `false` otherwise.
+ */
+inline bool bitfield_bit32_read(uint32_t bitfield,
+                                bitfield_bit32_index_t bit_index) {
+  return (bitfield & (0x1u << bit_index)) == (0x1u << bit_index);
+}
+
+/**
+ * Writes `value` to the `bit_index`th bit in `bitfield`.
+ *
+ * @param bitfield Bitfield to update the bit in.
+ * @param bit_index Bit to update.
+ * @param value Bit value to write to `bitfield`.
+ * @return `bitfield` with the `bit_index`th bit set to `value`.
+ */
+inline uint32_t bitfield_bit32_write(uint32_t bitfield,
```

Comment:
I prefer `set/clear` routines, where it becomes a simple case of  `bitfield |= (0x1 << bit_index)` and `bitfield &= ~(0x1 << bit_index)`.

Reply:
At some point, with separate set/clear routines, you end up using an if-statement to choose which one to do, where you could just pass the bool directly - or else you end up doing the `field32_write` with a single-bit mask, as `dif_i2c_write_byte_raw` was doing. It seems to me better to provide the generic version, and people can call it with a constant argument if they only need set/clear functionality.

Reply:
> At some point, with separate set/clear routines, you end up using an if-statement to choose which one to do

But most of the time you don't need to do that, for example when you need to set separate bits of the register during configuration, you will probably just need to call N set bit functions in a row. It probably mostly just the enable/disable API that needs an if, even so - it is simple enough to at the call time.

Reply:
I think the convnience of never having to write an `if`, versus sometimes writing a hard-coded true or false, is much greater than the readability gain of set/clear.

Reply:
Not a big deal, I guess it's just my preference. Marking this as resolved.